backend/cellguide/pipeline/gpt_descriptions/gpt_description_generator.py

```diff
@@ -0,0 +1,56 @@
+import logging
+import os
+
+from backend.cellguide.pipeline.config import CellGuideConfig
+from backend.cellguide.pipeline.providers.openai_provider import OpenAIProvider
+from backend.common.providers.s3_provider import S3Provider
+
+logger = logging.getLogger(__name__)
+
+
+def generate_new_gpt_descriptions(all_cell_type_ids_in_corpus: list[str]) -> dict[str, str]:
+    """
+    Generates GPT descriptions for all cell type ID's in the corpus if they haven't
+    already been generated. Cell type ids should be stored in the cellguide-data-public-<env>
+    bucket in the gpt_descriptions folder. For each cell type ID, we'll have a JSON file
+    that contains the GPT description
+
+    📁 gpt_descriptions
+    ﹂CL_0000000.json
+    ﹂CL_0000001.json
+    ﹂...
+    """
+    bucket_name = CellGuideConfig().bucket
+    new_gpt_descriptions: dict[str, str] = {}
+
+    for cell_type_id in all_cell_type_ids_in_corpus:
+        object_key = f"gpt_descriptions/{cell_type_id}.json"
+        if not S3Provider.does_object_exist(bucket_name=bucket_name, object_key=object_key):
+            logger.info(f"GPT description does not exist yet for {cell_type_id}, generating...")
+            description_str = OpenAIProvider.generate_chatgpt4_output(cell_type_id=cell_type_id)
+
+            new_gpt_descriptions[cell_type_id] = description_str
+            logger.info(f"Generated GPT description for {cell_type_id}: {description_str}")
+
+
+def upload_descriptions_to_s3(new_gpt_descriptions: dict[str, str]) -> None:
+    bucket_name = CellGuideConfig().bucket
+
+    for cell_type_id, description_str in new_gpt_descriptions.items():
+        object_key = f"gpt_descriptions/{cell_type_id}.json"
+
+        # Write to a local temp file
+        description_file = "/tmp/{cell_type_id}.json"
```

Comment:
You can use the context manager instead of writing to `/tmp/`... Sometimes `/tmp/` might not be writeable (at least in my experience).

```
from tempfile import TemporaryDirectory
with TemporaryDirectory() as output_dir:
.....
```

Reply:
updated - thanks for the pointer, didn't know about that package!